# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/wr/WebRenderCanvasRenderer.cpp
# Commit: bbfe23c61add
# Full Hash: bbfe23c61add07377daf80696fd6996c0463f5b6
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-06-09 09:21:34
# Regressor Bug: 1632249
# File Overlap Count: 1
# Description:
#   Bug 1632249 - Support compositing out-of-process WebGL. r=handyman,lsalzman,nical,geckoview-reviewers,agi,imanol
#   
#   * Majorly simplity CanvasRenderer
#   * Replace GLScreenBuffer with trivial GLSwapChain
#   * Use descriptor structs so that future SharedSurface changes aren't so painful
# ==============================================================================

diff -r 8fcfac450ecc -r bbfe23c61add gfx/layers/wr/WebRenderCanvasRenderer.cpp
--- a/gfx/layers/wr/WebRenderCanvasRenderer.cpp	Mon Jun 08 20:13:01 2020 +0000
+++ b/gfx/layers/wr/WebRenderCanvasRenderer.cpp	Mon Jun 08 20:34:15 2020 +0000
@@ -20,14 +20,14 @@
   return mManager->WrBridge();
 }
 
-void WebRenderCanvasRenderer::Initialize(const CanvasInitializeData& aData) {
-  ShareableCanvasRenderer::Initialize(aData);
+WebRenderCanvasRendererAsync::~WebRenderCanvasRendererAsync() {
+  if (mPipelineId.isSome()) {
+    mManager->RemovePipelineIdForCompositable(mPipelineId.ref());
+    mPipelineId.reset();
+  }
 }
 
-WebRenderCanvasRendererAsync::~WebRenderCanvasRendererAsync() { Destroy(); }
-
-void WebRenderCanvasRendererAsync::Initialize(
-    const CanvasInitializeData& aData) {
+void WebRenderCanvasRendererAsync::Initialize(const CanvasRendererData& aData) {
   WebRenderCanvasRenderer::Initialize(aData);
 
   ClearCachedResources();
@@ -35,25 +35,12 @@
 
 bool WebRenderCanvasRendererAsync::CreateCompositable() {
   if (!mCanvasClient) {
-    TextureFlags flags = TextureFlags::DEFAULT;
-    if (mOriginPos == gl::OriginPos::BottomLeft) {
-      flags |= TextureFlags::ORIGIN_BOTTOM_LEFT;
-    }
-
-    if (IsOpaque()) {
-      flags |= TextureFlags::IS_OPAQUE;
+    auto compositableFlags = TextureFlags::NO_FLAGS;
+    if (!mData.mIsAlphaPremult) {
+      // WR needs this flag marked on the compositable, not just the texture.
+      compositableFlags |= TextureFlags::NON_PREMULTIPLIED;
     }
-
-    if (!mIsAlphaPremultiplied) {
-      flags |= TextureFlags::NON_PREMULTIPLIED;
-    }
-
-    mCanvasClient = CanvasClient::CreateCanvasClient(GetCanvasClientType(),
-                                                     GetForwarder(), flags);
-    if (!mCanvasClient) {
-      return false;
-    }
-
+    mCanvasClient = new CanvasClient(GetForwarder(), compositableFlags);
     mCanvasClient->Connect();
   }
 
@@ -75,13 +62,6 @@
   }
 }
 
-void WebRenderCanvasRendererAsync::Destroy() {
-  if (mPipelineId.isSome()) {
-    mManager->RemovePipelineIdForCompositable(mPipelineId.ref());
-    mPipelineId.reset();
-  }
-}
-
 void WebRenderCanvasRendererAsync::
     UpdateCompositableClientForEmptyTransaction() {
   bool wasDirty = IsDirty();